[10.x] Refactored LazyCollection::except() to save memory

fuwasegu opened the pull request:
## Overview
I've reduced the memory usage by reimplementing the except() method of LazyCollection::class to use lazy evaluation.

## Motivation
LazyCollection is a sibling of the Collection class that utilizes Generators for lazy evaluation.
I believe that methods implemented by LazyCollection should be lazily evaluated using yield wherever possible.

However, while it's often unavoidable to use passthru() to delegate to the Collection class, doing so defeats the purpose of using LazyCollection. This is because all the data gets loaded into memory, negating the benefits of lazy evaluation.

## Tests
As this is a refactoring, the fact that no test modifications were needed indicates that the functionality remains unbroken.

## References
I previously refactored LazyCollection::take() for similar reasons in PR #48382.

## Performance Metrics
I measured the performance before and after the refactoring using the following test code, focusing on both execution time and memory usage. Please use these metrics to evaluate the effectiveness of this refactoring.

### Code
```php
<?php

declare(strict_types=1);

use Illuminate\Support\Enumerable;
use Illuminate\Support\Facades\Artisan;
use Illuminate\Support\LazyCollection;

Artisan::command('except:before {size}', function () {
    $factory = function (): Generator {
        $size = $this->argument('size');

        for ($index = 0; $index < $size; $index++) {
            yield "key-{$index}" => "value-{$index}";
        }
    };

    $collection = new LazyCollection($factory);
    $value = $collection->except(array_map(fn (int $i) => 'key-' . (10 ** $i), range(0, 8)))->first();

    dump(
        \memory_get_usage(true),
        \microtime(true) - \LARAVEL_START,
    );
});

Artisan::command('except:after {size}', function () {
    $factory = function (): Generator {
        $size = $this->argument('size');

        for ($index = 0; $index < $size; $index++) {
            yield "key-{$index}" => "value-{$index}";
        }
    };

    $collection = new class ($factory) extends LazyCollection {
        public function except($keys)
        {
            $keys = match (true) {
                $keys instanceof Enumerable => $keys->all(),
                is_array($keys) => $keys,
                is_null($keys) => [],
                default => func_get_args(),
            };

            if (count($keys) === 0) {
                return new static(function () {
                    yield from $this;
                });
            }

            return new static(function () use ($keys) {
                foreach ($this as $key => $value) {
                    if (!in_array($key, $keys)) {
                        yield $key => $value;
                    }
                }
            });
        }
    };
    $value = $collection->except(array_map(fn (int $i) => 'key-' . (10 ** $i), range(0, 8)))->first();

    dump(
        \memory_get_usage(true),
        \microtime(true) - \LARAVEL_START,
    );
});
```

### Results
About memory usage:
||100|10000|1000000|100000000|
|---|---|---|---|---|
|before|23068672|23068672|Fatal error|Fatal error|
|after|23068672|23068672|23068672|23068672|

About execution time:
||100|10000|1000000|100000000|
|---|---|---|---|---|
|before|0.058702945709229|0.062233924865723|Fatal error|Fatal error|
|after|0.059131860733032|0.058793067932129|0.05843710899353|0.059433937072754|

fuwasegu committed:
refactor: except method to lazy

fuwasegu committed:
formtting

JosephSilber reviewed (commented):
Add a test ensuring it's actually lazy. Update this `testExceptIsLazy` method:

https://github.com/laravel/framework/blob/ac91af7d25962d873dbd20edc1603969ffa42619/tests/Support/SupportLazyCollectionIsLazyTest.php#L345-L354

Add something like this between those 2 tests:

```php
$this->assertEnumerates(4, function ($collection) {
    $collection->except([1, 2])->take(2)->all();
});
```

fuwasegu commented on src/Illuminate/Collections/LazyCollection.php:
```diff
@@ -419,7 +419,26 @@ public function duplicatesStrict($callback = null)
+            default => func_get_args(),
+        };
+
+        if (count($keys) === 0) {
+            return new static(function () {
+                yield from $this;
+            });
+        }
```
This is where I wasn't sure if this was the right way.
Thanks!

fuwasegu committed:
Update src/Illuminate/Collections/LazyCollection.php

Co-authored-by: Joseph Silber <contact@josephsilber.com>

fuwasegu commented on src/Illuminate/Collections/LazyCollection.php:
```diff
@@ -419,7 +419,26 @@ public function duplicatesStrict($callback = null)
+
+        return new static(function () use ($keys) {
+            foreach ($this as $key => $value) {
+                if (! in_array($key, $keys)) {
```
i see.
I had not been able to take into account the performance of PHP down to the function level. Thanks for the advice！
I will try to fix it.

fuwasegu replied:
done!

fuwasegu committed:
test: lazy test

fuwasegu committed:
modify: use array_key_exists instead of in_array

fuwasegu committed:
test: fix test

fuwasegu committed:
formmating

driesvints commented:
@fuwasegu tests are fixed on 10.x, can you rebase?

fuwasegu commented:
@driesvints 
thank you for fixing tests.
i rebased and all tests were passed !

taylorotwell commented:
I think this may be a breaking change? Normal collections use `Arr::forget` under the hood for except, which supports "dot" notation.

JosephSilber commented:
> I think this may be a breaking change? Normal collections use `Arr::forget` under the hood for except, which supports "dot" notation.

Oh right!

When this PR came in, I was racking my brain trying to remember why I didn't do it originally.

Now that you've mentioned it, I remember that this was indeed the reason. Didn't think at the time that it's worth reimplementing it all here.

---

If we do want to do it now, we have to decide whether we want to support dot notation for when the children (or other descendants) are also lazy collections 🤔

fuwasegu commented:
@taylorotwell @JosephSilber 
thank you for your comment!

Certainly, the Normal Collection supports nested access by dot.
Taylor is right, my changes are disruptive.

The choices are
- Make this change from 11.x and allow destructive changes.
- Give up being lazy and passthru as before
- Support Nested access by dot

And Joseph is right, LazyCollection<LazyCollection> does not support nested access even in its current state, so we will have to think about whether to support it.

fuwasegu commented:
I've thought about it a lot, but I couldn't come up with an implementation of except() that supports dot nesting in LazyCollection...

For example, if the keys are ['foo.bar', 'bar', 'foo.baz'], etc., LazyCollection cannot run backwards, so it will have to go around once for each key verification

fuwasegu commented:
@morloderex  nice idea !
thanks!

fuwasegu commented:
How about this one?

```php
public function except($keys)
{
    $keys = match (true) {
        $keys instanceof Enumerable => $keys->all(),
        is_array($keys) => $keys,
        is_null($keys) => [],
        default => func_get_args(),
    };

    if (count($keys) === 0) {
        return $this;
    }
    
    $keys = Arr::undot(array_fill_keys($keys, true));

    $forget = function (LazyCollection|ArrayAccess|array &$data, array $keys) use (&$forget) {
        if ($data instanceof LazyCollection) {
            return new static(function () use ($data, $keys, $forget) {
                foreach ($data as $key => $value) {
                    $shouldYield = true;
                    foreach ($keys as $k => $v) {
                        if ($k === $key && is_array($v)) {
                            yield $key => $forget($value, $v);
                            $shouldYield = false;
                            break;
                        } elseif ($k === $key) {
                            $shouldYield = false;
                            break;
                        }
                    }
                    if ($shouldYield) {
                        yield $key => $value;
                    }
                }
            });
        } else {
            foreach ($keys as $k => $v) {
                if (is_array($v)) {
                    $forget($data, $v);
                } else {
                    if (isset($data[$k])) {
                        unset($data[$k]);
                    }
                }
            }
        }
    };

    return new static(function () use ($keys, $forget) {
        yield from $forget($this, $keys);
    });
}
```

fuwasegu commented:
How about this one ?

```php
public function except($keys)
{
    $keys = match (true) {
        $keys instanceof Enumerable => $keys->all(),
        is_array($keys) => $keys,
        is_null($keys) => [],
        default => func_get_args(),
    };

    if (count($keys) === 0) {
        return $this;
    }

    $keys = Arr::undot(array_fill_keys($keys, true));

    $forget = function (LazyCollection|ArrayAccess|array &$data, array $keys) use (&$forget) {
        if ($data instanceof LazyCollection) {
            return new static(function () use ($data, $keys, $forget) {
                foreach ($data as $key => $value) {
                    $shouldYield = true;
                    foreach ($keys as $k => $v) {
                        if (
                            $k === $key
                            && is_array($v)
                            && (
                                $value instanceof LazyCollection
                                || is_array($value)
                                || $value instanceof ArrayAccess
                            )
                        ) {
                            yield $key => $forget($value, $v);
                            $shouldYield = false;
                            break;
                        } elseif ($k === $key) {
                            $shouldYield = false;
                            break;
                        }
                    }
                    if ($shouldYield) {
                        yield $key => $value;
                    }
                }
            });
        } else {
            foreach ($keys as $k => $v) {
                if (is_array($v)) {
                    return $forget($data, $v);
                } else {
                    if (isset($data[$k])) {
                        unset($data[$k]);
                        return $data;
                    }
                }
            }
        }
    };

    return new static(function () use ($keys, $forget) {
        yield from $forget($this, $keys);
    });
}
```

fuwasegu commented:
this test cases was passed.

```php
<?php

namespace Tests\Unit;

use Illuminate\Support\Collection;
use Illuminate\Support\LazyCollection;
use PHPUnit\Framework\TestCase;

class ExceptTest extends TestCase
{

    public function test_1(): void
    {
        $collection = TestLazyCollection::make([1, 2, 3, 4]);
        $collection = $collection->except([0, 2]);
        $this->assertSame([1 => 2, 3 => 4], $collection->toArray());
    }

    public function test_2(): void
    {
        $collection = TestLazyCollection::make([
            'foo' => 'FOO',
            'bar' => [
                'one' => 'ONE',
                'two' => 'TWO',
            ],
            'baz' => 'BAZ',
        ]);
        $collection = $collection->except(['foo', 'bar.two']);
        $this->assertSame([
            'bar' => [
                'one' => 'ONE',
            ],
            'baz' => 'BAZ',
        ], $collection->toArray());
    }

    public function test_3(): void
    {
        $collection = TestLazyCollection::make([
            'foo' => 'FOO',
            'bar' => Collection::make([
                'one' => 'ONE',
                'two' => 'TWO',
            ]),
            'baz' => 'BAZ',
        ]);
        $collection = $collection->except(['foo', 'bar.two']);

        $this->assertEquals([
            'bar' => Collection::make([
                'one' => 'ONE',
            ]),
            'baz' => 'BAZ',
        ], iterator_to_array($collection));
    }

    public function test_4(): void
    {
        $collection = TestLazyCollection::make(function () {
            yield 'foo' => LazyCollection::make(function () {
                yield 'one' => 'ONE';
                yield 'two' => 'TWO';
                yield 'three' => [
                    'first' => 'FIRST',
                    'second' => 'SECOND',
                ];
            });
            yield 'bar' => LazyCollection::make(function () {
                yield 'one' => 'ONE';
                yield 'two' => 'TWO';
                yield 'three' => [
                    'first' => 'FIRST',
                    'second' => 'SECOND',
                ];
            });
            yield 'baz' => LazyCollection::make(function () {
                yield 'one' => 'ONE';
                yield 'two' => 'TWO';
                yield 'three' => [
                    'first' => 'FIRST',
                    'second' => 'SECOND',
                ];
            });
        });
        $collection = $collection->except(['foo', 'bar.two', 'baz.three.second']);

        $this->assertEquals([
            'bar' => [
                'one' => 'ONE',
                'three' => [
                    'first' => 'FIRST',
                    'second' => 'SECOND',
                ],
            ],
            'baz' => [
                'one' => 'ONE',
                'two' => 'TWO',
                'three' => [
                    'first' => 'FIRST',
                ],
            ],
        ], $collection->toArray());
    }

}

```

mpyw commented:
@fuwasegu Refactored and fixed some bugs

```php
public function except($keys)
{
    $keys = match (true) {
        $keys instanceof Enumerable => $keys->all(),
        is_array($keys) => $keys,
        is_null($keys) => [],
        default => func_get_args(),
    };

    if (count($keys) === 0) {
        return $this;
    }

    $keys = Arr::undot(array_fill_keys($keys, true));

    $forget = function (LazyCollection|ArrayAccess|array &$data, array $keys) use (&$forget) {
        if ($data instanceof LazyCollection) {
            return new static(function () use ($data, $keys, $forget) {
                foreach ($data as $key => $value) {
                    foreach ($keys as $k => $v) {
                        if (
                            $k === $key
                            && is_array($v)
                            && (
                                $value instanceof LazyCollection
                                || is_array($value)
                                || $value instanceof ArrayAccess
                            )
                        ) {
                            yield $key => $forget($value, $v);
                            continue 2;
                        }
                        if ($k === $key) {
                            continue 2;
                        }
                    }
                    yield $key => $value;
                }
            });
        }
        foreach ($keys as $k => $v) {
            if (is_array($v)) {
                $forget($data, $v);
            } else {
                unset($data[$k]);
            }
        }
    };

    return new static(function () use ($keys, $forget) {
        yield from $forget($this, $keys);
    });
}
```

fuwasegu commented:
For some reason, my previous comment was lost, so I'm writing again.

As taylor and joseph have said, this PR is destructive.
I didn't realize at first that except supports dot nesting and implemented it under the assumption of a flat LazyCollection.

Now that joseph says "we have to decide what to do if the child of a LazyCollection is a LazyCollection", I think we have three choices.
- Allow destructive changes and do not support nesting in the case of a LazyCollection. I would include this change starting with 11.x.
- Make LazyCollection::except() support nesting by dots
- LazyCollection::except() remains as before, allowing all data to be in memory with the call (giving up on being lazy)

Personally, I don't want to give up on the third one.
And I implemented except() for dots.
However, it has become a bit complicated, so I leave it to the owner to decide whether to include it as part of the framework implementation.

fuwasegu committed:
refactor: support dot notation

fuwasegu committed:
add: commnet

fuwasegu committed:
formmating

fuwasegu commented:
@taylorotwell @JosephSilber 

To test the changes accommodating dot notation, I ran tests in SupportCollectionTest::testExcept() using the following data:

```php
$data = new $collection([
    'social' => [
        'twitter' => '@taylorotwell',
        'github' => 'taylorotwell',
    ],
    'projects' => new $collection([
        'laravel' => [
            'stars' => 30000,
            'forks' => 5000,
        ],
        'spark' => [
            'stars' => 2000,
            'forks' => 500,
        ],
    ]),
]);

$this->assertEquals([
    'social' => [
        'github' => 'taylorotwell',
    ],
    'projects' => [
        'laravel' => [],
        'spark' => [
            'stars' => 2000,
        ],
    ],
], $data->except(['social.twitter', 'projects.laravel.stars', 'projects.laravel.forks', 'projects.spark.forks'])->toArray());
```

For LazyCollection, the test passed. However, for the standard Collection, I encountered the following warning, which caused the test to fail:

https://github.com/laravel/framework/blob/4b22e39a058a6e91418bf080c923eb6021295287/src/Illuminate/Collections/Arr.php#L286-L294

Is this because Collection::except() is not designed to accommodate types like Collection<TKey, Collection<TKey, array>>?

If so, the first step is to decide whether to allow such types. Should I consider refactoring Collection::except() with its own implementation, rather than relying on Arr::except()?

Angel5a commented:
@fuwasegu What about a compromise:
- Step1. If no dot notation involved, then use your simple lazy implementation, otherwise give up and fallback to old implementation (10.x branch).
- Step 2. Add support for dot notation (perhaps in 11.x branch).

Lazy collections that are not lazy is a bad thing. It would be nice to see it fixed.

fuwasegu commented:
@Angel5a  Thank you for your excellent suggestions!

Indeed.
- If $keys contains a string with a dot in it, passthru, and 
- Otherwise, use my first implementation and process lazily

is a very clear-cut compromise!

Dot support is already implemented, so it is possible to incorporate it starting with 11.x.

@taylorotwell @JosephSilber  What do you think of this idea?

JosephSilber commented:
Not sure about that. Leads to unintuitive swings in memory usage. Adding a dot now completely changes the memory footprint.

By selecting "fewer" values, I'll be increasing my memory footprint. Not what I'd expect.

Angel5a commented:
@JosephSilber Yes, that's the problem, but
1. We are **already have** the same issue.
2. Refactoring is a step by step process.
 
For now we get closer for desired behavior (negotiate specific use-case). For 11.x we can completely resolve the problem, even  by introducing BC change.

Penalty is in dot presence check.

taylorotwell commented:
I honestly just can't commit to maintaining that much code for this use case.
